### ISSUE TITLE
Action/Bean separation

### ISSUE THREAD

**Issue by jccarrillo** [StripesFramework/stripes#44]
All,

Is there any reason behind having an ActionBean instead of an Action and a Bean? I see some benefits on such separation.

**Comment by vankeisb** [comment]
Hiya,

I don't understand... Could you elaborate ? What's the use case ?

**Comment by jccarrillo** [comment]
For instance:
```
public class MyActionBean implements ActionBean {
    private String myString; // with a number of more attributes
    ... setters/getters
    @DefaultHandler
    public Resolution doSomething() {
        //
    }
}
```
Why not have the following?
```
public class MyBean {
    private String myString; // with a number of more attributes
    ... setters/getters
}
public class MyAction implements Action {
    @DefaultHandler
    public Resolution doSomething(MyBean myBean) {
        //
    }
}
```

This is just an observation.

JC

**Comment by vankeisb** [comment]
Well that's a design choice that was made from the very beginning. What you describe seems more like Spring or Play where controllers are methods (even static) that accept parameters bound from the request.
Both designs have pros and cons in terms of code factorization and reuse. 
I like Stripes', maybe because I'm used to it, but I think the "stateful" (bean+props vs static method + args) is easier to code and read. 
Also, it helps a lot with JSPs or other views, which just need the action as an entry point to all bound and computed props. Not to mention the taglib which is consistent because of this 'stateful' design choice.
So like all design choices there is always an alternative but I've always found the Stripes way to be effective (DRY, easy to understand validation logic etc). It all fits quite well doesn't it ? Do you think of any limitation in this model ?

**Comment by iluvtr** [comment]
Hi, JC, normally you don't put all the properties in the ActionBean, take a
look:

public class MyBean {
    private String myString; // with a number of more attributes
    ... setters/getters
}


public class MyActionBean implements ActionBean {

    @DefaultHandler
    public Resolution doSomething() {
        //
    }
}




2016-02-09 14:55 GMT-05:00 JC Carrillo <notifications@github.com>:

**Comment by iluvtr** [comment]
Hi, JC, normally you don't put all the properties in the ActionBean, take a
look:

public class MyBean {
    private String myString; // with a number of more attributes
    ... setters/getters
}


public class MyActionBean implements ActionBean {
    private MyBean myBean;

    @DefaultHandler
    public Resolution doSomething() {
        //do something with mybean
    }

     ... setters/getters


}





2016-02-10 8:04 GMT-05:00 Nestor Hernandez <iluvtr@gmail.com>:

**Comment by jccarrillo** [comment]
@vankeisb/@iluvtr,

Thank you for your responses!

Perhaps this design choice is what sets it apart from other frameworks.

I come from a Spring background and I'm more adjusted to Controllers rather than ActionBeans.
I like the concept of the Actionable Bean but I question the processing power, if any, that the JVM spends on creating these ActionBeans per request. If no other framework is used, I'd say it's minimal but if, for instance, you use a custom `ActionResolver.Class` to create your ActionBeans and wire/inject objects into them, such process 'may' take longer, I haven't done any testing to confirm so.
In comparison to Spring, the controllers are singletons with "pre-wired" instances and the beans are passed to them. Stripes doesn't have the concept of 'Injection', to my knowledge, but you can certainly override the `ActionResolver.Class` and control the creation and injection of the ActionBeans.

In short, my concern is with the 'custom' creation of the ActionBeans. I would expect the singleton approach to be more effective than the non-singleton/stripes approach. (note: no metrics have been done, this is speculation)

JC

**Comment by iluvtr** [comment]
Hi JC, IMHO Stripes' choice of creating actionBeans per request naturally
fits the way web works and you're free of thread safety troubles. There's
no perfomance penalty of creating a new ActionBeans because they're
lighweight instances.

2016-02-10 9:02 GMT-05:00 JC Carrillo <notifications@github.com>:

**Comment by vankeisb** [comment]
That's a legitimate question. 

Stripes resolves ActionBean *classes* very fast. The "map" of classes for a given url binding is computed once at startup, and cached. I don't know how fast Spring is in that regard (I mean the resolver), but Stripes is pretty good I think. 

Now as you said, it does create an instance of the bean class for every incoming http request. 

It has an overhead in theory, but in practice, I think you cannot even measure this with a profiler. It simply does Class<ActionBean>.newInstance() so it's about nothing compared to what you appserver has already done when it passes you the HttpServlet* objects...

As @iluvtr said, this is very lightweight, so I don't think there is a performance issue there. 

HTH

**Comment by timothystone** [comment]
Adding to @vankeisb in regards to @jccarrillo speed: I can speak as the View-Controller app architect for the acquisitions and servicing platforms of barclaycardus.com, both on Stripes+Tomcat. The servicing platform serves 16MM customers (with logins exceeding 100K/day) and credit applications exceeding 50K/day. Gomez currently ranks the applications as #1 for speed (granted there are a number of external FEO best practices lifting that number, but Stripes makes it possible): http://www.dynatrace.com/en/benchmarks/united-states/credit-card/ns

Sidebar: the effectiveness of the ActionBean pattern in onboarding devs also cannot be understated.

**Comment by jccarrillo** [comment]
All,

Thanks for all the feedback. Everything is clear now. Closing this issue.

JC
